### PR TITLE
Alerts Map

### PR DESCRIPTION
**Pull request by aholachek**
Hi, here's a quick summary of the major changes:
- The map is created in the Template.map.rendered function, and subscribes to Reports to listen for changes
- I gave the map tab its own route which required me to split the station lists out into their own templates, which can be swapped with the map from within the landing template. (Maybe could have been done more simply? This was my first time using Meteor and I'm not quite sure I understand the best way to do routing).
- The map updates are throttled to once every 15 seconds. 
- Silver line, mattapan, and commuter rail are currently unsupported due to mismatch between the mbta-ninja metadata and my stops.geojson file (and no geo data in case of the commuter rail).
- I added two new libraries: d3 (visualization) and leaflet (underlying map). 
- The alert triangles are added when the station has at least one report; they are pointed in roughly the relevant directions, and the user can click on them to open a tooltip that offers more information and a link to the relevant page(s).
- Just to recap, the reason I think a map would be a good idea is that people will want to know roughly where geographically the reports are occurring in the context of a single line e.g. reports on the Braintree end of the red line, for instance, probably won't impact someone in Davis (at least not immediately...), and the user shouldn't have to click through just to learn that the report is irrelevant in his/her case. In addition many people use multiple lines and it's better not to force them to click through to each line. Finally, in list form some of the lines are a bit overwhelming and hard to parse quickly (e.g. the green line dropdown) and, assuming there are fewer than 10 or 15 reports at any one time, I think it's faster to just glance at the map and see where the reports are clustered.

Please let me know what needs to be changed or improved. I'm happy to work more on it to get it ready for release, this was a fun project that has already taught me a lot about meteor and maps!

### CODE REVIEW

**Comment by radhikamalik** [comment]
Hey @aholachek thanks for submitting the PR! To start, could you please rebase and squash all 3 commits into one and also attach a screenshot of the app as it looks now? Thanks!

**Comment by aholachek** [comment]
Here's a sample screenshot:

screen shot 2015-11-05 at 7 02 28 pm

I rebased the commits into just the one.

**Comment by JacobEvelyn** [comment]
This is really cool!

File: client/index.html [hunk 1]
```diff
@@ -40,8 +40,8 @@
     ga('send', 'pageview');
   </script>
 
-  <script type="text/javascript">
+<script type="text/javascript">
```
Comment (radhikamalik) [inline client/index.html:43]: Why is this indentation changed? The change seems incorrect.

File: client/styles/style.css [hunk 2]
```diff
@@ -302,3 +302,98 @@ iframe#twitter-widget-0 {
...
+  width: 100%;
+  height: 100%:
+}
+
+.landing-tabs li.tab:hover {
+  border-bottom: 3px solid #DCDCDC;
+}
+
+.landing-tabs li.tab.active {
+border-bottom: 3px solid #039be5;
```
Comment (radhikamalik) [inline client/styles/style.css:336]: indent here

File: client/templates/landing.html [hunk 3]
```diff
@@ -14,21 +14,12 @@ <h1 class="app-title">
...
-
-      <h1 class="transport-type">Commuter rail</h1>
-      <ul class="collapsible" data-collapsible="accordion">
-        {{#each trainLines}}
-          {{> line}}
-        {{/each}}
-      </ul>
-
-      <div class="share-buttons">
+    {{ #if showMapView }}
```
Comment (radhikamalik) [inline client/templates/landing.html:17]: indentation again

File: client/styles/style.css [hunk 4]
```diff
@@ -302,3 +302,98 @@ iframe#twitter-widget-0 {
...
+
+#map .terminator-label {
+  text-anchor: middle;
+  font-size: 1.3em;
+  fill: #ABA9A9;
+  font-weight: bold;
+}
+
+#map .direction-arrow {
+fill: #fff;
```
Comment (radhikamalik) [inline client/styles/style.css:380]: fix indentation here.

File: client/styles/style.css [hunk 5]
```diff
@@ -302,3 +302,98 @@ iframe#twitter-widget-0 {
...
 }
+
+
+
+/* ==========================================================================
+   For the map visualization
+   ========================================================================== */
+
+/* tabs */
+ .landing-tabs i {
```
Comment (radhikamalik) [inline client/styles/style.css:313]: indentation here

File: client/templates/map.html [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<template name="map">
+
+  <ul class="landing-tabs">
+      <li class="tab"><a href="{{pathFor route='landing'}}"> <i class="mdi-action-list small"></i> List View</a></li>
+      <li class="tab active" ><a href="{{pathFor route='map'}}"><i class="mdi-maps-map small"></i> Map View</a></li>
+  </ul>
+
+<div>
```
Comment (radhikamalik) [inline client/templates/map.html:8]: indentation

File: client/templates/map.js [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+    blue: "#041299",
+    purple: "#801174",
+    silver: "#949494"
+  };
+
+  function chooseColor(d) {
+    var data = d.data ? d.data : d;
+    var line = data.properties.lines[0];
+    if (line == "mattapan") {
+      line = "red"
```
Comment (radhikamalik) [inline client/templates/map.js:26]: semicolon at the end of this line

File: client/templates/map.js [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+  var colors = {
+    red: "#C60813",
+    orange: "#FA6820",
+    green: "#0A6536",
+    blue: "#041299",
+    purple: "#801174",
+    silver: "#949494"
+  };
+
+  function chooseColor(d) {
```
Comment (radhikamalik) [inline client/templates/map.js:22]: can you add a comment here about what this function expects as argument and what it returns?

File: client/templates/map.js [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+    return colors[line];
+  }
+
+  var layer = L.tileLayer(
+    'http://{s}.basemaps.cartocdn.com/light_nolabels/{z}/{x}/{y}.png', {
+      attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a> &copy; <a href="http://cartodb.com/attributions">CartoDB</a>',
+      subdomains: 'abcd'
+    });
+
+  var northEast = L.latLng(42.5, -70.9),
```
Comment (radhikamalik) [inline client/templates/map.js:37]: these would be better declared as constants somewhere

File: client/templates/map.js [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+    var data = d.data ? d.data : d;
+    var line = data.properties.lines[0];
+    if (line == "mattapan") {
+      line = "red"
+    }
+    return colors[line];
+  }
+
+  var layer = L.tileLayer(
+    'http://{s}.basemaps.cartocdn.com/light_nolabels/{z}/{x}/{y}.png', {
```
Comment (radhikamalik) [inline client/templates/map.js:32]: please add a comment about what these URLs are referring to.

File: client/templates/map.js [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+    });
+  }
+
+  var force = d3.layout.force()
+    .charge(0)
+    .gravity(0);
+
+    //needs to be accessible across updateMap functions
+    var labelIntervalId;
+
```
Comment (radhikamalik) [inline client/templates/map.js:118]: extra line

File: client/templates/stationList.html [hunk 12]
```diff
@@ -0,0 +1,21 @@
+<template name="stationList">
+      <ul class="landing-tabs">
```
Comment (radhikamalik) [inline client/templates/stationList.html:2]: please make indentation consistent

File: client/templates/stationList.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
+
+Template.stationList.rendered = function(){
+
+  // Collapsible lines
+$('.collapsible').collapsible({
```
Comment (radhikamalik) [inline client/templates/stationList.js:5]: fix indentation

File: client/templates/map.js [hunk 14]
```diff
@@ -0,0 +1,545 @@
...
+    }
+
+    function gravity(k) {
+      return function(d) {
+        d.x += (d.x0 - d.x) * k;
+        d.y += (d.y0 - d.y) * k;
+      };
+    }
+
+    function collide(k) {
```
Comment (radhikamalik) [inline client/templates/map.js:460]: can you add a comment about what this function is meant for?

**Comment by radhikamalik** [comment]
This looks really neat!

I added a few stylistic comments on the code so we ensure it's consistent with the existing code.

I wonder if it makes sense to add a line somewhere (maybe on the map view page?) that this view only supports the main lines (no silver line or commuter rail lines).

Also looking at it on mobile, looks like the map starts out with some unnecessary space at the top above "Oak Grove" (and you have scroll down to see all the lines covered completely). It would be good to get rid of this space at the top so you can see all lines together on the map at a glance.
screen shot 2015-11-14 at 5 42 24 pm

Thoughts?

**Comment by radhikamalik** [comment]
One more thing: we might want to make the buttons to add the reports common to both views so someone doesn't have to necessarily go back to the list view to create a report.

**Comment by aholachek** [comment]
Hi Radhika,
I'm sorry that it took so long for me to reply! Thanks for your comments and suggested fixes. Before I get started making the changes, and maybe a few cosmetic changes to the station markers that I've been thinking about, I just want to confirm that this is a feature that should be added to the project. I went to one of the Tuesday meetups and talked to one of the guys who is a leader of the group and map expert (I've unfortunately forgotten his name); he didn't think the map was necessary. Personally I still think it would be useful, for the reasons I wrote above, but I don't want to force through a feature that has limited utility. I'd love to hear your thoughts.

**Comment by modestmaya** [comment]
@patrickgreenwell do you know if (at this point in time) there would be any plans to move ahead with the addition of the map feature? Seems like it could be a great feature!
